### PR TITLE
NDEV-2475 collect metrics for requests to Solana

### PR DESCRIPTION
**Pull request by andrei-neon**

### CODE REVIEW

**Comment by andrei-neon** [comment]
fvffvfv

**Comment by andrei-neon** [comment]
gjrbkjre

**Comment by andrei-neon** [comment]
|              |   getHealth |   getVersion |    getBalance |   getLatestBlockhash |   requestAirdrop |   getSignatureStatuses |   isBlockhashValid |   getAccountInfo |   getTokenAccountBalance |   sendTransaction |   getFirstAvailableBlock |        getSlot |      getBlocks |       getBlock |   getMinimumBalanceForRentExemption |   simulateTransaction |   getTransaction |   getBlockTime |   getMultipleAccounts |   getSignaturesForAddress |   getClusterNodes |   unknown |
|:-------------|------------:|-------------:|--------------:|---------------------:|-----------------:|-----------------------:|-------------------:|-----------------:|-------------------------:|------------------:|-------------------------:|---------------:|---------------:|---------------:|------------------------------------:|----------------------:|-----------------:|---------------:|----------------------:|--------------------------:|------------------:|----------:|
| count        | 3           | 532          | 1497          |        3739          |        32        |          1022          |       493          |   43620          |                        3 |    18612          |                        2 | 26033          | 7305           | 3706           |                                   1 |          23           |   20312          | 4792           |         6899          |                  8        |             1     |     1     |
| average_time | 0.000333333 |   0.00105075 |    0.00131463 |           0.00215619 |         0.000625 |             0.00102544 |         0.00043002 |       0.00138556 |                        0 |        0.00362105 |                        0 |     0.00112288 |    0.000528405 |    0.000992715 |                                   0 |           0.000434783 |       0.00386476 |    0.000616235 |            0.00227395 |                  0.000875 |             0.105 |     0.013 |
| max_time     | 0.001       |   0.015      |    0.034      |           0.104      |         0.001    |             0.093      |         0.015      |       0.099      |                        0 |        0.389      |                        0 |     0.103      |    0.044       |    0.025       |                                   0 |           0.001       |       0.389      |    0.095       |            0.139      |                  0.004    |             0.105 |     0.013 |
| min_time     | 0           |   0          |    0          |           0          |         0        |             0          |         0          |       0          |                        0 |        0          |                        0 |     0          |    0           |    0           |                                   0 |           0           |       0          |    0           |            0          |                  0        |             0.105 |     0.013 |
| median_time  | 0           |   0          |    0          |           0.001      |         0.001    |             0          |         0          |       0          |                        0 |        0.002      |                        0 |     0          |    0           |    0.001       |                                   0 |           0           |       0.001      |    0           |            0.001      |                  0.0005   |             0.105 |     0.013 |

**Comment by andrei-neon** [comment]
|   getHealth |   getVersion |    getBalance |   getLatestBlockhash |   requestAirdrop |   getSignatureStatuses |   isBlockhashValid |   getAccountInfo |   getTokenAccountBalance |   sendTransaction |   getFirstAvailableBlock |        getSlot |      getBlocks |       getBlock |   getMinimumBalanceForRentExemption |   simulateTransaction |   getTransaction |   getBlockTime |   getMultipleAccounts |   getSignaturesForAddress |   getClusterNodes |   unknown |
|------------:|-------------:|--------------:|---------------------:|-----------------:|-----------------------:|-------------------:|-----------------:|-------------------------:|------------------:|-------------------------:|---------------:|---------------:|---------------:|------------------------------------:|----------------------:|-----------------:|---------------:|----------------------:|--------------------------:|------------------:|----------:|
| 3           | 532          | 1497          |        3739          |        32        |          1022          |       493          |   43620          |                        3 |    18612          |                        2 | 26033          | 7305           | 3706           |                                   1 |          23           |   20312          | 4792           |         6899          |                  8        |             1     |     1     |
| 0.000333333 |   0.00105075 |    0.00131463 |           0.00215619 |         0.000625 |             0.00102544 |         0.00043002 |       0.00138556 |                        0 |        0.00362105 |                        0 |     0.00112288 |    0.000528405 |    0.000992715 |                                   0 |           0.000434783 |       0.00386476 |    0.000616235 |            0.00227395 |                  0.000875 |             0.105 |     0.013 |
| 0.001       |   0.015      |    0.034      |           0.104      |         0.001    |             0.093      |         0.015      |       0.099      |                        0 |        0.389      |                        0 |     0.103      |    0.044       |    0.025       |                                   0 |           0.001       |       0.389      |    0.095       |            0.139      |                  0.004    |             0.105 |     0.013 |
| 0           |   0          |    0          |           0          |         0        |             0          |         0          |       0          |                        0 |        0          |                        0 |     0          |    0           |    0           |                                   0 |           0           |       0          |    0           |            0          |                  0        |             0.105 |     0.013 |
| 0           |   0          |    0          |           0.001      |         0.001    |             0          |         0          |       0          |                        0 |        0.002      |                        0 |     0          |    0           |    0.001       |                                   0 |           0           |       0.001      |    0           |            0.001      |                  0.0005   |             0.105 |     0.013 |

**Comment by andrei-neon** [comment]
|                                   |   count |   average_time |   max_time |   min_time |   median_time |
|:----------------------------------|--------:|---------------:|-----------:|-----------:|--------------:|
| getHealth                         |       3 |    0.000333333 |      0.001 |      0     |        0      |
| getVersion                        |     532 |    0.00105075  |      0.015 |      0     |        0      |
| getBalance                        |    1497 |    0.00131463  |      0.034 |      0     |        0      |
| getLatestBlockhash                |    3739 |    0.00215619  |      0.104 |      0     |        0.001  |
| requestAirdrop                    |      32 |    0.000625    |      0.001 |      0     |        0.001  |
| getSignatureStatuses              |    1022 |    0.00102544  |      0.093 |      0     |        0      |
| isBlockhashValid                  |     493 |    0.00043002  |      0.015 |      0     |        0      |
| getAccountInfo                    |   43620 |    0.00138556  |      0.099 |      0     |        0      |
| getTokenAccountBalance            |       3 |    0           |      0     |      0     |        0      |
| sendTransaction                   |   18612 |    0.00362105  |      0.389 |      0     |        0.002  |
| getFirstAvailableBlock            |       2 |    0           |      0     |      0     |        0      |
| getSlot                           |   26033 |    0.00112288  |      0.103 |      0     |        0      |
| getBlocks                         |    7305 |    0.000528405 |      0.044 |      0     |        0      |
| getBlock                          |    3706 |    0.000992715 |      0.025 |      0     |        0.001  |
| getMinimumBalanceForRentExemption |       1 |    0           |      0     |      0     |        0      |
| simulateTransaction               |      23 |    0.000434783 |      0.001 |      0     |        0      |
| getTransaction                    |   20312 |    0.00386476  |      0.389 |      0     |        0.001  |
| getBlockTime                      |    4792 |    0.000616235 |      0.095 |      0     |        0      |
| getMultipleAccounts               |    6899 |    0.00227395  |      0.139 |      0     |        0.001  |
| getSignaturesForAddress           |       8 |    0.000875    |      0.004 |      0     |        0.0005 |
| getClusterNodes                   |       1 |    0.105       |      0.105 |      0.105 |        0.105  |
| unknown                           |       1 |    0.013       |      0.013 |      0.013 |        0.013  |

**Comment by andrei-neon** [comment]
|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getHealth                         |       3 |      0     |      0.001 |    0.000333333 |        0      |
| getVersion                        |     532 |      0     |      0.015 |    0.00105075  |        0      |
| getBalance                        |    1497 |      0     |      0.034 |    0.00131463  |        0      |
| getLatestBlockhash                |    3739 |      0     |      0.104 |    0.00215619  |        0.001  |
| requestAirdrop                    |      32 |      0     |      0.001 |    0.000625    |        0.001  |
| getSignatureStatuses              |    1022 |      0     |      0.093 |    0.00102544  |        0      |
| isBlockhashValid                  |     493 |      0     |      0.015 |    0.00043002  |        0      |
| getAccountInfo                    |   43620 |      0     |      0.099 |    0.00138556  |        0      |
| getTokenAccountBalance            |       3 |      0     |      0     |    0           |        0      |
| sendTransaction                   |   18612 |      0     |      0.389 |    0.00362105  |        0.002  |
| getFirstAvailableBlock            |       2 |      0     |      0     |    0           |        0      |
| getSlot                           |   26033 |      0     |      0.103 |    0.00112288  |        0      |
| getBlocks                         |    7305 |      0     |      0.044 |    0.000528405 |        0      |
| getBlock                          |    3706 |      0     |      0.025 |    0.000992715 |        0.001  |
| getMinimumBalanceForRentExemption |       1 |      0     |      0     |    0           |        0      |
| simulateTransaction               |      23 |      0     |      0.001 |    0.000434783 |        0      |
| getTransaction                    |   20312 |      0     |      0.389 |    0.00386476  |        0.001  |
| getBlockTime                      |    4792 |      0     |      0.095 |    0.000616235 |        0      |
| getMultipleAccounts               |    6899 |      0     |      0.139 |    0.00227395  |        0.001  |
| getSignaturesForAddress           |       8 |      0     |      0.004 |    0.000875    |        0.0005 |
| getClusterNodes                   |       1 |      0.105 |      0.105 |    0.105       |        0.105  |
| unknown                           |       1 |      0.013 |      0.013 |    0.013       |        0.013  |

File: .github/workflows/deploy.py [hunk 1]
```diff
@@ -293,7 +298,7 @@ def get_all_containers_logs():
 
     ssh_client.connect(hostname=proxy_ip, username='root',
                        key_filename=ssh_key, timeout=120)
-    services = ["postgres", "dbcreation", "indexer", "proxy", "faucet"]
+    services = ["postgres", "dbcreation", "indexer", "proxy", "faucet", "nginx"]
```
Comment (kristinaNikolaevaa) [inline .github/workflows/deploy.py]: nginx is run on other server , right?

File: .github/workflows/deploy.py [hunk 2]
```diff
@@ -467,5 +472,85 @@ def process_output(output):
...
+
+def calculate_stats(stats):
+    formated_stats = {key: {} for key in stats.keys()}
+    for method, data in stats.items():
+        formated_stats[method]["count"] = len(data["times"])
+        formated_stats[method]["average_time"] = statistics.mean(data["times"])
+        formated_stats[method]["max_time"] = max(data["times"])
+        formated_stats[method]["min_time"] = min(data["times"])
+        formated_stats[method]["median_time"] = statistics.median(data["times"])
+    return formated_stats
```
Comment (kristinaNikolaevaa) [inline .github/workflows/deploy.py]: let's sort it by count field

File: .github/workflows/deploy.py [hunk 3]
```diff
@@ -467,5 +472,85 @@ def process_output(output):
...
+            raise RuntimeError(f"Attempt to leave a comment on a PR failed: {response.text}")
+
+
+@cli.command("post_comment", help="Post comment to the PR")
+@click.option("--message", help="Message to post")
+@click.option("--pull_request", help="PR number")
+@click.option("--token", help="Github token")
+def post_comment(message, pull_request, token):
+    gh_client = GithubClient(token)
+    gh_client.add_comment_to_pr(message, pull_request)
```
Comment (kristinaNikolaevaa) [inline .github/workflows/deploy.py:572]: let's first delete comment if it exist (as for dapps report) 
To have only 1 comment if the build was restarted

File: .github/workflows/full_test_suite/solana_init.sh [hunk 4]
```diff
@@ -56,9 +56,28 @@ services:
...
 EOF
 
-
 # wake up Solana
 docker-compose -f docker-compose-ci.yml -f docker-compose-ci.override.yml pull solana
 docker-compose -f docker-compose-ci.yml -f docker-compose-ci.override.yml up -d solana
+
+docker-compose -f docker-compose-ci.yml -f docker-compose-ci.override.yml pull nginx
+docker-compose -f docker-compose-ci.yml -f docker-compose-ci.override.yml up -d nginx
+docker cp nginx.conf $(docker ps -qf "name=nginx"):/etc/nginx/nginx.conf
```
Comment (kristinaNikolaevaa) [inline .github/workflows/full_test_suite/solana_init.sh]: why you don't copy it through volumes in docker-compose ? (by this way you don't need restart the container)

Comment (kristinaNikolaevaa) [reply .github/workflows/full_test_suite/solana_init.sh]: also I think you can do pull and up at the same time with solana to save time
llike 
docker-compose -f docker-compose-ci.yml -f docker-compose-ci.override.yml pull nginx solana

File: .github/workflows/pipeline.yml [hunk 5]
```diff
@@ -355,6 +355,37 @@ jobs:
...
+    env:
+      SOLANA_IP: ${{ needs.prepare-infrastructure.outputs.solana_ip }}
+    steps:
+      - uses: actions/checkout@v4
+        with:
+           fetch-depth: 0
+      - name: Install pandas, tabulate
+        run: |
+          pip install pandas
+          pip install tabulate
```
Comment (kristinaNikolaevaa) [inline .github/workflows/pipeline.yml]: Ask DEVOPS team to install it once on test-runners

**Review by kristinaNikolaevaa — CHANGES_REQUESTED** [review]

File: .github/workflows/deploy.py (hunk 1 above)
Comment (andrei-neon) [reply .github/workflows/deploy.py]: Yes, correct, I'll remove it.

**Comment by andrei-neon** [comment]
|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getHealth                         |       3 |      0     |      0.009 |    0.003       |         0     |
| getVersion                        |     534 |      0     |      0.03  |    0.00181461  |         0.001 |
| getAccountInfo                    |   44164 |      0     |      0.137 |    0.0023621   |         0.001 |
| getBalance                        |    1528 |      0     |      0.131 |    0.00219045  |         0.001 |
| getLatestBlockhash                |    3795 |      0     |      0.138 |    0.00274809  |         0.001 |
| requestAirdrop                    |      32 |      0     |      0.004 |    0.0010625   |         0.001 |
| getSignatureStatuses              |    1022 |      0     |      0.128 |    0.00223581  |         0.001 |
| isBlockhashValid                  |     493 |      0     |      0.043 |    0.00115619  |         0     |
| getTokenAccountBalance            |       4 |      0     |      0.001 |    0.00075     |         0.001 |
| sendTransaction                   |   17167 |      0     |      0.146 |    0.00520965  |         0.003 |
| getFirstAvailableBlock            |       2 |      0.001 |      0.001 |    0.001       |         0.001 |
| getSlot                           |   25212 |      0     |      0.142 |    0.00180993  |         0.001 |
| getBlocks                         |    7289 |      0     |      0.055 |    0.000912196 |         0.001 |
| getBlock                          |    3702 |      0     |      0.042 |    0.001557    |         0.001 |
| getMinimumBalanceForRentExemption |       1 |      0     |      0     |    0           |         0     |
| simulateTransaction               |      23 |      0     |      0.001 |    0.000565217 |         0.001 |
| getTransaction                    |   18826 |      0     |      0.168 |    0.00573956  |         0.001 |
| getBlockTime                      |    4890 |      0     |      0.054 |    0.00114131  |         0     |
| getMultipleAccounts               |    7086 |      0     |      0.144 |    0.00342817  |         0.001 |
| getSignaturesForAddress           |       7 |      0     |      0.036 |    0.00742857  |         0.001 |
| getClusterNodes                   |       1 |      0.001 |      0.001 |    0.001       |         0.001 |

File: .github/workflows/deploy.py (hunk 2 above)
Comment (andrei-neon) [reply .github/workflows/deploy.py]: Done ✅

File: .github/workflows/deploy.py (hunk 3 above)
Comment (andrei-neon) [reply .github/workflows/deploy.py:572]: Done ✅

File: .github/workflows/full_test_suite/solana_init.sh (hunk 4 above)
Comment (andrei-neon) [reply .github/workflows/full_test_suite/solana_init.sh]: Fixed ✅

File: .github/workflows/pipeline.yml (hunk 5 above)
Comment (andrei-neon) [reply .github/workflows/pipeline.yml]: Done ✅

**Comment by neonlabstech** [comment]
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getSlot                           |   70227 |          0 |      0.085 |    0.000680166 |        0      |
| getAccountInfo                    |   54196 |          0 |      0.085 |    0.0013518   |        0      |
| getBlocks                         |   28022 |          0 |      0.043 |    0.000476733 |        0      |
| getTransaction                    |   22852 |          0 |      0.1   |    0.00319486  |        0.001  |
| sendTransaction                   |   20790 |          0 |      0.085 |    0.0032229   |        0.002  |
| getBlock                          |   14073 |          0 |      0.029 |    0.00086165  |        0.001  |
| getMultipleAccounts               |    9181 |          0 |      0.079 |    0.00190273  |        0.001  |
| getBlockTime                      |    5865 |          0 |      0.03  |    0.000600171 |        0      |
| getLatestBlockhash                |    4454 |          0 |      0.091 |    0.0016648   |        0.001  |
| getBalance                        |    1750 |          0 |      0.053 |    0.00166171  |        0      |
| getSignatureStatuses              |    1046 |          0 |      0.023 |    0.000891013 |        0      |
| getVersion                        |     544 |          0 |      0.074 |    0.00115441  |        0      |
| isBlockhashValid                  |     505 |          0 |      0.023 |    0.000360396 |        0      |
| getSignaturesForAddress           |     136 |          0 |      0.005 |    0.00182353  |        0.002  |
| requestAirdrop                    |      32 |          0 |      0.001 |    0.00065625  |        0.001  |
| simulateTransaction               |      23 |          0 |      0.001 |    0.000391304 |        0      |
| getHealth                         |       3 |          0 |      0.001 |    0.000666667 |        0.001  |
| getTokenAccountBalance            |       3 |          0 |      0.001 |    0.000666667 |        0.001  |
| getFirstAvailableBlock            |       2 |          0 |      0.001 |    0.0005      |        0.0005 |
| getMinimumBalanceForRentExemption |       1 |          0 |      0     |    0           |        0      |
| getClusterNodes                   |       1 |          0 |      0     |    0           |        0      |

</details>

File: .github/workflows/pipeline.yml [hunk 6]
```diff
@@ -355,6 +355,46 @@ jobs:
...
+    needs:
+      - prepare-infrastructure
+      - basic-tests
+      - build-image
+      - dapps-tests
+      - openzeppelin-tests
+    if: |
+      always() && 
+      contains(fromJSON('["success", "cancelled", "failure", "skipped", "completed"]'), needs.openzeppelin-tests.result) &&
+      contains(fromJSON('["success", "cancelled", "failure", "skipped", "completed"]'), needs.basic-tests.result) &&
```
Comment (kristinaNikolaevaa) [inline .github/workflows/pipeline.yml]: basic tests can't be skipped

I think always() is aanough here if you all statuses are good for the job

Comment (andrei-neon) [reply .github/workflows/pipeline.yml]: Makes sense. I have removed check for the basic tests 👍

**Comment by neonlabstech** [comment]
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getSlot                           |   70134 |      0     |      0.183 |    0.00114997  |         0.001 |
| getAccountInfo                    |   54768 |      0     |      0.209 |    0.0021395   |         0.001 |
| getBlocks                         |   28495 |      0     |      0.077 |    0.00071967  |         0.001 |
| getTransaction                    |   19266 |      0     |      0.185 |    0.00629119  |         0.001 |
| sendTransaction                   |   17200 |      0     |      0.19  |    0.00587715  |         0.003 |
| getBlock                          |   14340 |      0     |      0.08  |    0.00137357  |         0.001 |
| getMultipleAccounts               |    9678 |      0     |      0.214 |    0.00315003  |         0.001 |
| getBlockTime                      |    6081 |      0     |      0.116 |    0.000970564 |         0     |
| getLatestBlockhash                |    4476 |      0     |      0.191 |    0.00274441  |         0.001 |
| getBalance                        |    1759 |      0     |      0.044 |    0.00199602  |         0.001 |
| getSignatureStatuses              |    1033 |      0     |      0.125 |    0.00190319  |         0.001 |
| getVersion                        |     538 |      0     |      0.04  |    0.0017342   |         0.001 |
| isBlockhashValid                  |     498 |      0     |      0.024 |    0.000696787 |         0     |
| getSignaturesForAddress           |     138 |      0     |      0.034 |    0.00449275  |         0.003 |
| requestAirdrop                    |      32 |      0.001 |      0.002 |    0.0011875   |         0.001 |
| simulateTransaction               |      23 |      0     |      0.001 |    0.000565217 |         0.001 |
| getHealth                         |       3 |      0.001 |      0.018 |    0.00666667  |         0.001 |
| getTokenAccountBalance            |       3 |      0     |      0.001 |    0.000333333 |         0     |
| getFirstAvailableBlock            |       2 |      0.001 |      0.001 |    0.001       |         0.001 |
| getMinimumBalanceForRentExemption |       1 |      0.001 |      0.001 |    0.001       |         0.001 |
| getClusterNodes                   |       1 |      0.001 |      0.001 |    0.001       |         0.001 |

</details>

**Comment by neonlabstech** [comment]
<details><summary>Solana Requests Statistics</summary>



|                                   |   count |   min_time |   max_time |   average_time |   median_time |
|:----------------------------------|--------:|-----------:|-----------:|---------------:|--------------:|
| getSlot                           |   75073 |      0     |      0.123 |    0.000752841 |         0     |
| getAccountInfo                    |   54207 |      0     |      0.147 |    0.00143911  |         0     |
| getBlocks                         |   29544 |      0     |      0.081 |    0.000470248 |         0     |
| getTransaction                    |   22543 |      0     |      0.156 |    0.00401393  |         0.001 |
| sendTransaction                   |   20484 |      0     |      0.089 |    0.00375932  |         0.002 |
| getBlock                          |   14828 |      0     |      0.036 |    0.000893175 |         0.001 |
| getMultipleAccounts               |    9364 |      0     |      0.148 |    0.00225075  |         0.001 |
| getBlockTime                      |    5903 |      0     |      0.113 |    0.000651703 |         0     |
| getLatestBlockhash                |    4448 |      0     |      0.091 |    0.00240355  |         0.001 |
| getBalance                        |    1751 |      0     |      0.037 |    0.00151913  |         0     |
| getSignatureStatuses              |    1012 |      0     |      0.073 |    0.00110178  |         0     |
| getVersion                        |     527 |      0     |      0.03  |    0.00106452  |         0     |
| isBlockhashValid                  |     488 |      0     |      0.014 |    0.000405738 |         0     |
| getSignaturesForAddress           |     149 |      0     |      0.007 |    0.00209396  |         0.002 |
| simulateTransaction               |      41 |      0     |      0.001 |    0.000585366 |         0.001 |
| requestAirdrop                    |      32 |      0     |      0.001 |    0.000625    |         0.001 |
| getHealth                         |       3 |      0     |      0.001 |    0.000333333 |         0     |
| getTokenAccountBalance            |       3 |      0     |      0.001 |    0.000333333 |         0     |
| getFirstAvailableBlock            |       2 |      0.001 |      0.001 |    0.001       |         0.001 |
| getMinimumBalanceForRentExemption |       1 |      0.001 |      0.001 |    0.001       |         0.001 |
| getClusterNodes                   |       1 |      0.001 |      0.001 |    0.001       |         0.001 |

</details>

**Review by kristinaNikolaevaa — APPROVED** [review]

**Pull request merged**